### ISSUE TITLE
CSS Modules failing when used in conjuction with ExtractTextWebpackPlugin

### ISSUE THREAD

**Issue by iamdustan** [webpack/webpack#3095]
I’ve been working on upgrading our app to webpack 2 and everything worked swimmingly for a while.

Before @TheLarkInn pointed me to http://webpack.js.org/how-to/upgrade-from-webpack-1/ we were 90% working with the old `module: {loaders: [...]}` format using `css-loader?modules` and the ExtractTextWebpackPlugin.

The 10% was configuring our stylus plugin with autoprefixer since polluting the top level object is no longer allowed.

After upgrading to the new `module {rules: [...]}` format I get module parse errors of various forms.

https://github.com/iamdustan/webpack-2-extract-css-modules-repro has a minimal repro case (without the stylus configuration at this point, but will add that in once I get through correcting the initial setup)

**Comment by TheLarkInn** [comment]
@jhnns do you think you'd be able to help me with this one. If we need to move it to `webpack/css-loader` or `webpack/extract-text-webpack-plugin` we can.

**Comment by iamdustan** [comment]
Some snippets:

**Using `{loader: css-loader: options: {...}}`**

```
const getCSSLoaders = (env) => {
  if (env === 'development') {
    return [
      { loader: 'style-loader' },
      { loader: 'css-loader',
        options: {modules: true, localIdentName: '[local]__[hash:base64:5]'},
      },
    ];
  }

  return extractCSS.extract([
    { loader: 'css-loader',
      options: {modules: true, localIdentName: '[hash:base64:5]'},
    },
  ]);
};
```

inline: ok.
extracted:

```
ERROR in ./app/styles.css
Module build failed: ModuleParseError: Module parse failed: /Users/dkasten/projects/webpack-2/app/styles.css Unexpected token (2:0)
You may need an appropriate loader to handle this file type.
|
| .base {
|   margin: 1.5em;
|   padding: 0;
    at /Users/dkasten/projects/webpack-2/node_modules/webpack/lib/NormalModule.js:207:34
    ....
```

-----

**Using `{loader: css-loader: query: {...}}`**

Dev: ok
Prod:

```
ERROR in ./app/styles.css
Module build failed: SyntaxError: Syntax error at line 1 column 33 of the JSON5 data. Still to read: "!css-loader?{\"module"
    at JSON5.parse.error (/Users/dkasten/projects/webpack-2/node_modules/loader-utils/node_modules/json5/lib/json5.js:56:25)
    ....
```

**Comment by iamdustan** [comment]
Some progress: https://github.com/iamdustan/webpack-2-extract-css-modules-repro/commit/adec7250b92f90b4117118458df206cfed283507

tldr; using the webpack 1 module resolve format (`loader` instead of `use` and `query` instead of `options`) gets the world working.

Now onto adding Stylus into the mix and seeing if I can keep it working with the other tools. 💪

**Comment by jhnns** [comment]
I've cloned the repo and everything is working. Is this resolved? What was the issue? Do we need to add something to the docs?

**Comment by richardscarrott** [comment]
I too am seeing a similar issue:

```
Module build failed: SyntaxError: Syntax error at line 1 column 26 of the JSON5 data. Still to read: "!style-loader!css-lo"
```

With the following config:

```
    const localCSS = ExtractTextPlugin.extract({
        fallbackLoader: 'style-loader',
        loader: {
            loader: 'css-loader',
            // Having to use `query` instead of `options`
            // https://github.com/webpack/extract-text-webpack-plugin/issues/265
            // https://github.com/webpack/extract-text-webpack-plugin/blob/master/index.js#L132
            query: {
                modules: true
            }
        }
    });

    return [{
        test: /\.css$/,
        use: [localCSS, 'autoprefixer-loader']
    }];
```

**Comment by iamdustan** [comment]
@jhnns just got back from vacation. HEAD of that repo is working because of this commit: https://github.com/iamdustan/webpack-2-extract-css-modules-repro/commit/adec7250b92f90b4117118458df206cfed283507#diff-11e9f7f953edc64ba14b0cc350ae7b9dR30

This is the unexpected part (which may have more real issues I have yet to surface or stumble upon).

The core issue seems to be that the new loader format of `use` and `options` keys do not work in conjunction with the ExtractTextPlugin. With the disclaimer of I have no idea what I’m talking about I would suspect that ExtractTextPlugin needs to be updated to be aware of and resolve the new loader format.

**Comment by jhnns** [comment]
Yes, unfortunately the extract-text-webpack-plugin hasn't caught up with the new loader configuration of webpack 2. We need to solve that before releasing webpack 2.

We continue at https://github.com/webpack/extract-text-webpack-plugin/issues/265

**Comment by iamdustan** [comment]
:thumbsup: thanks @jhnns.
